File: BinaryTree.py
class Node:
    def __init__(self):
        self.value = None
        self.left = None
        self.right = None

class BinaryTree:
    def __init__(self):
        self.__root = None
        self.__value_found = None

    def __insert(self,value, tree):
            if value < tree.value:
                if tree.left is None:
                    tree.left = Node()
                    tree.left.value = value
                else:
                    self.__insert(value,tree.left)
            if value >= tree.value:
                if tree.right is None:
                    tree.right = Node()
                    tree.right.value = value
                else:
                    self.__insert(value,tree.right)


    def insert(self, value):
        if self.__root is None:
            self.__root = Node()
            self.__root.value = value
        else:
            self.__insert(value,self.__root)


    def __bfs(self,value,tree):
        queue = []
        leaf = Node()
        queue.append(tree)

        while (queue != []):
            leaf = queue.pop(0)
            if leaf.value == value:
                self.__value_found = True
                return

            if leaf.left != None:
                queue.append(leaf.left)
            if leaf.right != None:
                queue.append(leaf.right)

        return

    def bfs(self, value):
        if self.__root.value == value:
            return True
        else:
            self.__value_found = False
            self.__bfs(value, self.__root)
            return self.__value_found

    def __show(self, tree):
        leaf = Node()
        queue = []
        queue.append(tree)

        while(queue != []):
            leaf = queue.pop(0)
            print leaf.value
            
            if leaf.left is not None:
                queue.append(leaf.left)
            if leaf.right is not None:
                queue.append(leaf.right)

    def show(self):
        self.__show(self.__root)

    def __dfs(self,value,tree, found = False):
        if self.__value_found == True:
            return
        if tree == None:
            return
        if tree.value == value:
            self.__value_found = True
            return
        else:
            self.__dfs(value,tree.left)
            self.__dfs(value,tree.right)
        return

    def dfs(self, value):
        if self.__root.value == value:
            return True
        else:
            self.__value_found = False
            self.__dfs(value, self.__root)
            return self.__value_found
        

def main():
    A = BinaryTree()
    A.insert(4)
    A.insert(5)
    A.insert(3)
    A.insert(6)
    A.insert(2)
    A.show()
    print A.bfs(9)
    print A.bfs(3)
    print A.dfs(3)
    print A.dfs(7)

if __name__ == "__main__":
    main()

File: BinaryTree.cpp
#include <iostream>
#include <queue>

//A simple node structure for the tree
struct Node {
  int value;
  Node* left;
  Node* right;
  bool visited;
};

//A simple binary tree
class BinaryTree {
public:
  //Constructs a root node filled with NULL values
  BinaryTree();
  //Eliminates the whole tree using the erase_tree function
  ~BinaryTree();
  //Inserts a node traversing in dfs mode.
  void insert(int value);
  //Performs Depth First Search. Checks if given value is in tree
  bool dfs(int value);
  //Performs Breadth First Search. Checks if given value is in tree
  bool bfs(int value);
  //Prints tree node values while traversing it in BFS mode
  void show();
private:
  //Deletes nodes of the tree from the bottom up traversing in DFS mode
  void erase_tree(Node* tree);
  void insert(int value, Node*& tree);
  void dfs(int value, Node* tree);
  bool bfs(int value, Node* tree);
  bool value_found;
  Node* root;
};

BinaryTree::BinaryTree(){
  root = new Node;
  root->value = NULL;
  root->left = NULL;
  root->left = NULL;
  root->visited = false;
  
}

BinaryTree::~BinaryTree(){
  erase_tree(root);
}


void BinaryTree::insert(int value, Node*& tree){


  if (tree == NULL){ 
    tree = new Node;          
    tree->value = value;
    tree->left = NULL;
    tree->right = NULL;
    tree->visited = false;

    return;

  }

  if(value < tree->value){ 
    insert(value,tree->left);
  }else if (value > tree->value){
    insert(value,tree->right);
  }
  

}

void BinaryTree::insert(int value){

  if(root->value == NULL){ 
    root->value = value;
  }else{ 
    insert(value,root);
  }
  return;
}

//DFS starting from the left
void BinaryTree::dfs(int value, Node* tree){
  static bool found;

  if (value_found == true) {
    return;
  }

  if(tree == NULL){
    return;
  }

  if(tree->visited == true){
    return;
  }

  if(value == tree->value){
    value_found = true;
    return;
  }else{
    dfs(value,tree->left);
    dfs(value,tree->right);
  }

  return;
}

bool BinaryTree::dfs(int value){
  value_found = false;

  if(root->value == value){
    value_found = true;
    return value_found;
  }
  dfs(value, root);
  return value_found;
}

bool BinaryTree::bfs(int value, Node* tree){
  Node* leaf;
  std::queue<Node *> queue;
  queue.push(tree);           
  
  while(!queue.empty()){         

    leaf = queue.front();         //visit node
    queue.pop();
    if(value == leaf->value){    //found => true
      return true;
    }
                                //push same level nodes
    if(leaf->left != NULL){      
      queue.push(leaf->left);
    }

    if(leaf->right != NULL){
      queue.push(leaf->right);
    }
    
    
  }
  return false;
}

void BinaryTree::show(){
  Node* leaf;
  std::queue<Node*> queue;
  queue.push(root);

  while(!queue.empty()){

    leaf = queue.front();
    std::cout << leaf->value << " ";  //print node
    queue.pop();

    
    if(leaf->left != NULL){     //traverse in bfs style
      queue.push(leaf->left);
    }

    if(leaf->right != NULL){
      queue.push(leaf->right);
    }

  }
  std::cout << std::endl;
  return;
}

bool BinaryTree::bfs(int value){
  if(root->value == value){         //base case
    return true;
  }
  
  bfs(value, root);
}

void BinaryTree::erase_tree(Node* tree){
  if(tree->value == NULL){
    return;
  }
  if(tree->left != NULL){
    erase_tree(tree->left);
  }
  if(tree->right != NULL){
    erase_tree(tree->right);
  }

  delete tree;   //erase from the bottom up
  return;
}


int main(){  //Testing the Class
  BinaryTree A;
  A.insert(5);
  A.insert(3);
  A.insert(4);
  A.insert(8);
  A.insert(9);
  A.show();
  std::cout << std::endl;
  if(A.dfs(10) == true){
    std::cout << "Found" << std::endl;
  }else{
    std::cout << "Not Found" << std::endl;
  }

  return 0;
}
